### DISCUSSION TITLE
How to update an index without first dropping it in Milvus Standalone?

### DISCUSSION THREAD

**Discussion by shauryagoel** [milvus-io/milvus#25856] [Q&A and General discussion] [answered]
In Milvus Standalone, I have built an index for a collection and made a FastAPI endpoint that calls the `collection.search(...)`. I have made a global variable (`collection`) which loads the collection. Later, whenever I want to insert more data into the collection and update the index, I would have to first release the collection and then, drop the index. This actually makes the search endpoint fail as the index has been removed from the RAM. Is there a way to prevent this? I want to simultaneously update the index as well as keep using the previous index till the new index is not built and loaded.

**Comment by xiaofan-luan** [comment]
> 

No don't really need to do release collection every time you search.
Just do insertion would work

**Reply by shauryagoel** [reply]
I am only releasing the collection when I need to update the index (as is required by Milvus) not every time whenever I search.
If I only do insertion, then, the new data would not be included in the index, hence, slowing down the search. 
I wanted to ask if there is a way to not release the collection while updating an index? This is necessary as if we release the collection then, the searching would not be possible.

**Comment by xiaofan-luan** [comment]
> I am only releasing the collection when I need to update the index (as is required by Milvus) not every time whenever I search. If I only do insertion, then, the new data would not be included in the index, hence, slowing down the search. I wanted to ask if there is a way to not release the collection while updating an index? This is necessary as if we release the collection then, the searching would not be possible.

you don't need to release the colleciton, or reindex the data.
data will be automatically indexed once you insert

**Reply by shauryagoel** [reply]
I know that once I insert the new data, it would be reindexed, but, it will be reindexed only when the growing segment is sealed. I think calling `flush()` would seal the segment. 

What if I want to change the parameters of the index? For example, increasing `nlist` in IVF_FLAT, when more data is inserted.

**Comment by xiaofan-luan** [comment]
> I know that once I insert the new data, it would be reindexed, but, it will be reindexed only when the growing segment is sealed. I think calling `flush()` would seal the segment.
> 
> What if I want to change the parameters of the index? For example, increasing `nlist` in IVF_FLAT, when more data is inserted.

They you will need to release collection and drop index

**Reply by shauryagoel** [reply]
Isn't there any way to avoid this downtime? 
Would it be possible if I use Milvus cluster mode?

**Reply by xiaofan-luan** [reply]
If you have double memory resources.
Then what you can do is create new colleciton
insert, index and load.
Change the alias point to the new collection

**Reply by shauryagoel** [reply]
So you are saying that I would have to create a new collection with a different name? Then, insert the same data as in the previous collection and then create the new desired index using it. Finally, remove the already existing collection and then change the new collection's name to be the same as the original collection's name.

**Reply by yhmo** [reply, marked as answer]
Milvus doesn't allow multiple index on a vector field. So, if you want to change the parameter of the old index, you have drop the old index and create new one. But during recreating the index, the collection will not able to service for long time. To avoid this situation, you can prepare anther collection and use "[alias](https://milvus.io/docs/collection_alias.md)" to swap the two collection smoothly. The following steps show how to do that:
1.  assume you have a collection "C1" with index IVF_FLAT
2.  now you want to change the index to be HNSW
3.  assign a alias to collection "C1", let's say, the alias is "proxy_name"
```
from pymilvus import utility
utility.create_alias(
  collection_name = "C1",
  alias = "proxy_name"
)
```

4. modify your client code, use the "proxy_name" to replace all the "C1", now the server knows `Collection("proxy_name")` is equal to `Collection("C1")`
```
collection = Collection("proxy_name")
collection.search()
```

5. create a new collection "C2", insert the same data into this collection
6. build index HNSW for "C2"
7. once the index of "C2" is finished, assign the alias "proxy_name" to collection "C2"
```
from pymilvus import utility
utility.alter_alias(
  collection_name = "C2",
  alias = "proxy_name"
)
```

8. now, all the search() requests are sent to "C2", because the server knows "proxy_name" is "C2". You don't need to change any code of search
9. drop the collection "C1"
